compiler/rustc_mir_build/src/thir/cx/expr.rs

```diff
@@ -93,6 +94,20 @@ impl<'tcx> Cx<'tcx> {
     ) -> Expr<'tcx> {
         let Expr { temp_lifetime, .. } = expr;
 
+        // FIXME(generic_const_exprs): Once expressions of kind `AdressOf`,
+        // `Borrow` and `Deref` are supported, we can allow adjustments here
+        if let ExprKind::Literal {
```

Comment:
I would assume that `&*expr` is used for borrowck reasons :thinking: 

While I am not too familiar with this code, this impl seems worrying to me because:
- in places where we only have one of the adjustments, i.e. either `&CONST` or `*CONST` this is incorrect. Don't know if something like that is possible rn but it seems likely to me.
- this probably causes some subtle changes to borrowck, though considering that const params can currently only be `&'static Ty` this is probably not too much of a problem :shrug:

Reply:
I completely misunderstood the issue, I initially thought that we would reject these borrows and derefs anyway, and this seems to be the case inside anonymous const blocks when `generic_const_exprs` is  used, where we always throw an error when we encounter them during the construction of the `AbstractConst` ([here](https://github.com/rust-lang/rust/blob/18fa4342fc215fe9e714307db694eaa8f5dc4a0d/compiler/rustc_trait_selection/src/traits/const_evaluatable.rs#L403-L409)), but treating all Borrow and Deref exprs as nops on all constants in general is not what we want obviously. 

If we do want to allow reborrows, then the current implementation of raising the bug during the construction of the `AbstractConst` seems to actually be what we want here right? And we should just keep this until we allow Borrows and Derefs inside anonymous constant.

Reply:
Maybe we should try to give a better error message here that makes the implicit adjustments explicit to the user?